File: Python/textbook_track/chapter03/ba3e.py
# Construct the De Bruijn Graph of a Collection of k-mers
import sys

from collections import defaultdict


class DeBruijnGraph:
    def __init__(self, k_mers):
        self._adjacency_list = build_de_bruijn_graph(k_mers)

    def edges(self):
        for prefix, neighbours in self._adjacency_list.items():
            neighbour_string = ','.join(neighbours)
            yield f'{prefix} -> {neighbour_string}'


def build_de_bruijn_graph(k_mers):
    d = defaultdict(list)
    for k_mer in k_mers:
        prefix = k_mer[:-1]
        suffix = k_mer[1:]
        d[prefix].append(suffix)
    return dict(d)


def main():
    data = sys.stdin.read().splitlines()
    graph = DeBruijnGraph(data)
    for edge in graph.edges():
        print(edge)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter03/BA3E.scala
package TextbookTrack.Chapter03

object BA3E {
  class DeBruijnGraph(kMers: List[String]) {
    import DeBruijnGraph.buildDeBruijnGraph

    def edges: Iterator[String] =
      for {
        (kMer, neighbours) <- adjacencyList.iterator
        neighbourString: String = neighbours.mkString(",")
      } yield s"$kMer -> $neighbourString"

    private val adjacencyList: Map[String, List[String]] = buildDeBruijnGraph(kMers)
  }

  object DeBruijnGraph {
    private def buildDeBruijnGraph(kMers: List[String]): Map[String, List[String]] =
      kMers
        .map(kMer => (kMer.dropRight(1), kMer.drop(1)))
        .groupBy{ case (prefix, _) => prefix }
        .view
        .mapValues(_.map{ case (_, suffix) => suffix })
        .toMap
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val graph = new DeBruijnGraph(reader.toList)
    graph.edges.foreach(println)
  }
}
